### PR TITLE
wip : feat(pkg) Support creation of Openshift Routes

### PR DESCRIPTION
**Pull request by waveywaves**
# Changes

This WiP work for supporting Openshift Route API which would help users using Openshift to have routes crated when they use the Operator. This work is towards non interference in any other creation of resource, but only if the Openshift Route API exists then create it. 

- Checks if the Openshift API exists
- If it does then creates the Route Resource for the HTTP Service 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes tests (if functionality changed/added)
- [ ] Includes docs (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/jenkinsci/kubernetes-operator/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/jenkinsci/kubernetes-operator/blob/master/CONTRIBUTING.md) for more details._

## Reviewer Notes

If API changes are included, additive changes must be approved by at least two [OWNERS](https://github.com/jenkinsci/kubernetes-operator/blob/master/OWNERS) and backwards incompatible changes must be approved by [more than 50% of the OWNERS](https://github.com/jenkinsci/kubernetes-operator/blob/master/OWNERS).

# Release Notes

```
Describe any user facing changes here, or delete this block.
Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

### CODE REVIEW

**Comment by waveywaves** [comment]
@akram @tomaszsek PTAL

File: go.mod [hunk 1]
```diff
@@ -13,28 +13,28 @@ require (
...
 	github.com/go-logr/zapr v0.1.1
 	github.com/go-openapi/spec v0.19.4
-	github.com/gogo/protobuf v1.3.1
+	github.com/gogo/protobuf v1.3.1 // indirect
 	github.com/golang/mock v1.3.1
+	github.com/google/gofuzz v1.1.0 // indirect
 	github.com/mailgun/mailgun-go/v3 v3.6.0
-	github.com/operator-framework/operator-sdk v0.15.1
-	github.com/pborman/uuid v1.2.0
+	github.com/openshift/api v3.9.1-0.20190924102528-32369d4db2ad+incompatible
```
Comment (akram) [inline go.mod:20]: what happened with these?

Comment (akram) [reply go.mod:20]: and also this one? 
better to keep it untouched.

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 2]
```diff
@@ -173,24 +183,30 @@ func (r *ReconcileJenkinsBaseConfiguration) ensureResourcesRequiredForJenkinsPod
...
-	if err := r.createRBAC(metaObject); err != nil {
+	httpServiceName := resources.GetJenkinsHTTPServiceName(r.Configuration.Jenkins)
+	if err := r.createService(metaObject, httpServiceName, r.Configuration.Jenkins.Spec.Service); err != nil {
 		return err
 	}
-	r.logger.V(log.VDebug).Info("Service account, role and role binding are present")
+	r.logger.V(log.VDebug).Info("Jenkins HTTP Service is present")
 
-	if err := r.ensureExtraRBAC(metaObject); err != nil {
+	if err := r.createRoute(metaObject, httpServiceName); err != nil && routeAPIFound {
```
Comment (jkhelil) [inline pkg/controller/jenkins/configuration/base/reconcile.go:192]: if verifyRouteApi() && err := r.createRoute(metaObject, httpServiceName); err != nil

File: pkg/apis/apis.go [hunk 3]
```diff
@@ -11,7 +13,16 @@ var AddToSchemes runtime.SchemeBuilder
 
 // AddToScheme adds all Resources to the Scheme
 func AddToScheme(s *runtime.Scheme) error {
-	return AddToSchemes.AddToScheme(s)
+	err := AddToSchemes.AddToScheme(s)
+	if err != nil {
+		return errors.New("AddToSchemes Not added to scheme")
```
Comment (jkhelil) [inline pkg/apis/apis.go:18]: dont capitalize errors

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 4]
```diff
@@ -37,6 +42,10 @@ const (
 	fetchAllPlugins = 1
 )
 
+var (
+	routeAPIFound = false
```
Comment (jkhelil) [inline pkg/controller/jenkins/configuration/base/reconcile.go:46]: Is this needed ?

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 5]
```diff
@@ -1092,3 +1133,39 @@ func (r *ReconcileJenkinsBaseConfiguration) handleAdmissionControllerChanges(cur
...
+	// API Exists
+	return true, nil
+}
+
+func verifyRouteAPI() error {
+	found, err := verifyAPI(routev1.GroupName, routev1.SchemeGroupVersion.Version)
+	if err != nil {
+		return err
+	}
+	routeAPIFound = found
```
Comment (jkhelil) [inline pkg/controller/jenkins/configuration/base/reconcile.go:1169]: Is this needed?

File: pkg/apis/apis.go [hunk 6]
```diff
@@ -11,7 +13,16 @@ var AddToSchemes runtime.SchemeBuilder
...
+	}
+	err = routev1.AddToScheme(s)
```
Comment (akram) [inline pkg/apis/apis.go:20]: where are the other AddToSChemes for the other apis we use ?

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 4 above)
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:46]: better to externalize this in another file of package as it is not related specifically to the reconciler

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 7]
```diff
@@ -46,6 +55,7 @@ type ReconcileJenkinsBaseConfiguration struct {
 
 // New create structure which takes care of base configuration
 func New(config configuration.Configuration, logger logr.Logger, jenkinsAPIConnectionSettings jenkinsclient.JenkinsAPIConnectionSettings) *ReconcileJenkinsBaseConfiguration {
+	verifyRouteAPI()
```
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:58]: is there a way to run it during the init phase instead ?

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 8]
```diff
@@ -173,24 +183,30 @@ func (r *ReconcileJenkinsBaseConfiguration) ensureResourcesRequiredForJenkinsPod
 	}
 	r.logger.V(log.VDebug).Info("ConfigurationAsCode Secret and ConfigMap added watched labels")
 
-	if err := r.createRBAC(metaObject); err != nil {
+	httpServiceName := resources.GetJenkinsHTTPServiceName(r.Configuration.Jenkins)
+	if err := r.createService(metaObject, httpServiceName, r.Configuration.Jenkins.Spec.Service); err != nil {
 		return err
 	}
-	r.logger.V(log.VDebug).Info("Service account, role and role binding are present")
+	r.logger.V(log.VDebug).Info("Jenkins HTTP Service is present")
```
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:190]: why these changes ?

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 9]
```diff
@@ -344,7 +360,11 @@ func (r *ReconcileJenkinsBaseConfiguration) createServiceAccount(meta metav1.Obj
 	serviceAccount := &corev1.ServiceAccount{}
 	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: meta.Name, Namespace: meta.Namespace}, serviceAccount)
 	if err != nil && apierrors.IsNotFound(err) {
-		serviceAccount = resources.NewServiceAccount(meta, r.Configuration.Jenkins.Spec.ServiceAccount.Annotations)
+		annotations := r.Configuration.Jenkins.Spec.ServiceAccount.Annotations
+		if annotations == nil {
+			annotations = make(map[string]string)
```
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:365]: Shouldn't we add at this step the oauth annotation ?

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 10]
```diff
@@ -470,6 +491,26 @@ func (r *ReconcileJenkinsBaseConfiguration) createService(meta metav1.ObjectMeta
...
+	route := routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: meta.Namespace,
+			Labels:    meta.Labels,
+		},
+	}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: meta.Namespace}, &route)
+	if err != nil && apierrors.IsNotFound(err) {
+		resources.UpdateRouteFromService(&route, name)
```
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:504]: shouldn't we simply remove the route and recreate it if the Service changed ?

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 11]
```diff
@@ -1092,3 +1133,39 @@ func (r *ReconcileJenkinsBaseConfiguration) handleAdmissionControllerChanges(cur
 		}
 	}
 }
+
+func verifyAPI(group string, version string) (bool, error) {
```
Comment (akram) [inline pkg/controller/jenkins/configuration/base/reconcile.go:1137]: we should move this to another file or package.
and as it returns a boolean, rename it: isXYZ or hasXYZ

File: pkg/apis/apis.go (hunk 6 above)
Comment (waveywaves) [reply pkg/apis/apis.go:20]: They have been registered in `register.go`

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 2 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:192]: On every reconcilation we will be checking if the route api is available in this case

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 8 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:190]: reordering, this is from another PR I was working on

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 11 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:1137]: renamed it for now.

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 5 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:1169]: It will be used to check if the route api exists

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 9 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:365]: Removing this change currently as route should also be configurable from the CR if it is present.

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 7 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:58]: Any other place you can suggest, I found this to be convenient as it is checked when the reconciler config is set which only happens once.

File: pkg/apis/apis.go (hunk 3 above)
Comment (waveywaves) [reply pkg/apis/apis.go]: done

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 4 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go:46]: it will be used to read if the routeAPIExists. In the case of running this on Kubernetes, Operator might break.

File: go.mod (hunk 1 above)
Comment (waveywaves) [reply go.mod:19]: Updated, adding only openshift API modules

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 10 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: UpdatRouteFromService changed to PopulateRouteFromService, cuz we are not really updating here

**Comment by waveywaves** [comment]
Closing this in favour of ingress in the v1alpha3 version
